protocol/0030-ETHM-multisig_control_spec.md

```diff
@@ -8,45 +8,60 @@ Vega controls and maintains a number of Ethereum smart contracts which have func
-This means that the message to sign will look like this: 
-    `abi.encode(abi.encode(asset_source, asset_id, amount, address_of_user, nonce, "withdraw_asset"), bridge_address)`
-The resulting message is then hashed to a bytes32 message_hash which is then signed by a Vega validator (that has been previously added as a valid signer to the `MultisigControl` smart contract)
- 
-Each signature is a 65 byte hex string in the format `0x1230477...` the `0x` is removed (except for the first in the bundle) and appended to the end of the bundle of other validator signatures. 
-This bundle is the `signatures` parameter ultimately passed to `verify_signatures`. In the `verify_signatures` function this signature bundle is broken into individual signatures and checked against the valid signer list. If the number of valid signatures if above the threshold %, this is a valid bundle.   
-
-## Adding Signers
-Additional Vega validator nodes are added as MultisigControl signers by running the `add_signer` function.
-
-Following the information above from "Signature Bundles", the message to sign will look like this:
-`abi.encode(abi.encode(new_signer_address, nonce, "add_signer"), submitter)`
-In this case, the submitter will most likely be the new signer since they are effectively the proposer. The reason the submitter is not a contract as in the case with the bridges is that the function is called directly by a user, not by a smart contract, like the bridge. 
-
-## Removing Signer
-Vega validator nodes are removed from being MultisigControl signers by running the `remove_signer` function.
-Following the information above from "Signature Bundles", the message to sign will look like this:
-`abi.encode(abi.encode(old_signer_address, nonce, "remove_signer"), submitter)`
-In this case, the submitter will likely be the user/party who proposed the removal of that node as a signer.  
+In order for a smart contract to be securely controlled by the Vega network, it must reference a previously launched and configured `Multisig_Control` smart contract and call `verify_signatures` before executing a Vega controlled function.    
 
+A signature bundle is a hex string of appended 65 byte ECDSA signatures where each signer signed the same message hash usually containing the parameters of the function being called, the current signer set data consisting of a list of signers and their corresponding weights, and a nonce that is used to prevent replay attacks.
+
+## Signer Set Data
+Signer Set Data is the set of signer addresses and their associated weight. Once hashed, the signer set data hash must match the current signer set data hash stored on the smart contract. This hash can be updated by calling `update_signers` (see below). The entire set of addresses and weights needs to be sent with every verified transaction and will be invalid once the signer set (or weights) change.
+
+## Verify Signatures
```

Comment:
I think we should consider the signature format a bit more, both taking into context what Clef supports and also additional safe-guards.

I propose the following format:

```
concat(
  bytes2("\x19\x00"), bytes20(address(this)), // [1]
  u256(block.chainid), // [2]
  bytes32(current_signer_set_hash), // [3]
  bytes20(calling_contract), // [4]
  u256(nonce), // [5]
  bytes(function_payload) // [6]
)
```

1. This prefix is to accommodate Clef's [`text/validator` mime-type](http://adam.schmideg.net/go-ethereum/clef/apis#account_signdata). See their [source](https://github.com/ethereum/go-ethereum/blob/d36e6fc49d1ca3d7f76fb45d411369e1e6ee89d6/signer/core/signed_data.go#L201)
2. This prevents replay across chains for whatever reason. It's unlikely that the same signer set and multisig address will be present across chains, but this rules out that completely
3. This is the signer set discussed in the PR. I would also be beneficial if the signer set had a total-order such that when verifying signatures, the verification algorithm can be O(n) in the signers and not have to keep a set of seen signers like the current implementation. The order has to be easy to compute on the Vega side, so a simple lexicographic sort of the validator keys would be very easy. Or it could sort by weight in descending order.
4. Making the `calling_contract` explicit. In the current format we have the concept of `submitter` which has a dual-use; when the multisig contact is called by another contract, this becomes the address of the calling contact, however when it is the multisig contract verifying a call to itself, it becomes the address of whoever invokes the contact. In this case we should inject `address(this)` from the multisig contract itself and keep `msg.sender` for external contracts.
5. The nonce is a common-denominator control to prevent replay attacks since we cannot guarantee the order in which transactions to the multisig contract are done. This might as well be a part of the `function_payload`.
6. The function name and parameters as describe in the PR.

Reply:
[1] why do we care about clef? Is it giving us anything useful for the extra gas? 
[2] might consider doing this as it would allow validators to reuse addresses across multiple chains if we get there... A) they probably shouldn't do that and B) need to investigate the extra gas
[3] that hash is entirely derived from the passed in signer set rather than storing signers, this also includes the weights (good call and already implemented, will update the spec to make that clear)
[4] at the moment I have an internal validation function and an external one (see [source](https://github.com/vegaprotocol/Multisig_Control_V2/blob/main/contracts/Multisig_Control.sol)) Does this work?
[5] how common/feasable is sig recovery overlap for ETH addresses? (looking it up right after I post this)
[6] cool

Reply:
1. So in Vega you can either import an ethereum key or use a running clef service. Currently we have a fork of clef to support the current signature format: https://docs.vega.xyz/testnet/node-operators/how-to/rotate-ethereum-keys
2. Cost here is low. `CHAINID` is 2 gas, memory expansion and hashing additional bytes is 6 and 3 gas respectively.
3. Correct, my point was to define a total order, but as mentioned, it's something that should be defined elsewhere in the spec
4. Main point here is today, if you submit anything to the multisig contract itself, you need to use a fixed external account, while other contracts such as the bridge, can be submitted by anyone, since the "point of view" of who the submitter is changes. What I'm suggesting is that we make this the same so when transactions to the multisig contract itself are validated, we use the multisig address as the "submitter"
5. I don't remember what the point was here
6. :+1:

Reply:
1. I'll need to investigate clef more
2. seems reasonable, added to the spec on the end of the "final hash"
3. this was added to the spec as one of the assumptions
4. submitter is now always the contract calling the verification and never a user. even if it's on the multisig itself, this allows anyone to submit any of the multisig orders and removes that complication